### ISSUE TITLE
AttributeError: 'zoneinfo.ZoneInfo' object has no attribute 'localize'

### ISSUE THREAD

**Issue by deep-42-thought** [pimutils/khal#1087]
I have khal 0.10.4 installed on linux (crux) and encounter the above error when trying to set up a new event in ikhal.
I had this error before simply on displaying an event (e.g. opening ikhal), but I could get rid of it by adding
`local_timezone = Europe/Berlin` to my `.config/khal/config`.
Googling a little seems to point at a too-new tzlocal - I am indeed running on tzlocal 3.0

```
Traceback (most recent call last):
  File "/usr/bin/ikhal", line 33, in <module>
    sys.exit(load_entry_point('khal==0.10.4', 'console_scripts', 'ikhal')())
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/usr/lib/python3.9/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/lib/python3.9/site-packages/khal/cli.py", line 485, in interactive_cli
    controllers.interactive(
  File "/usr/lib/python3.9/site-packages/khal/controllers.py", line 545, in interactive
    pane = ui.ClassicView(
  File "/usr/lib/python3.9/site-packages/khal/ui/__init__.py", line 1045, in __init__
    daywalker = Walker(
  File "/usr/lib/python3.9/site-packages/khal/ui/__init__.py", line 361, in __init__
    self.ensure_date(this_date)
  File "/usr/lib/python3.9/site-packages/khal/ui/__init__.py", line 370, in ensure_date
    pile = self._get_events(day)
  File "/usr/lib/python3.9/site-packages/khal/ui/__init__.py", line 480, in _get_events
    self.events = sorted(self._collection.get_events_on(day))
  File "/usr/lib/python3.9/site-packages/khal/khalendar/khalendar.py", line 152, in get_events_on
    localize = self._locale['local_timezone'].localize
AttributeError: 'zoneinfo.ZoneInfo' object has no attribute 'localize'
```

regards, deep42thought

**Comment by deep-42-thought** [comment]
these changes are hacky, but make it work for me:
```
--- /usr/lib/python3.9/site-packages/khal/icalendar.py	2021-09-17 20:47:33.437285998 +0200
+++ /usr/lib/python3.9/site-packages/khal/icalendar.py	2021-09-17 20:41:18.280009757 +0200
@@ -101,8 +101,8 @@
         raise ValueError("no summary given")
 
     if not allday and timezone is not None:
-        dtstart = timezone.localize(dtstart)
-        dtend = timezone.localize(dtend)
+        dtstart = pytz.timezone(str(timezone)).localize(dtstart)
+        dtend =  pytz.timezone(str(timezone)).localize(dtend)
 
     event = icalendar.Event()
     event.add('dtstart', dtstart)
--- /usr/lib/python3.9/site-packages/khal/khalendar/khalendar.py	2021-09-17 20:47:33.590731428 +0200
+++ /usr/lib/python3.9/site-packages/khal/khalendar/khalendar.py	2021-09-17 20:40:19.623344156 +0200
@@ -31,6 +31,7 @@
 import os
 import os.path
 from typing import Any, Dict, Iterable, List, Optional, Tuple, Union  # noqa
+import pytz
 
 from . import backend
 from .event import Event
@@ -149,7 +150,7 @@
         start = dt.datetime.combine(day, dt.time.min)
         end = dt.datetime.combine(day, dt.time.max)
         floating_events = self.get_floating(start, end)
-        localize = self._locale['local_timezone'].localize
+        localize = pytz.timezone(str(self._locale['local_timezone'])).localize
         localized_events = self.get_localized(localize(start), localize(end))
         return itertools.chain(localized_events, floating_events)
 
```

**Comment by fire-space** [comment]
> I have khal 0.10.4 installed on linux (crux) and encounter the above error when trying to set up a new event in ikhal. I had this error before simply on displaying an event (e.g. opening ikhal), but I could get rid of it by adding `local_timezone = Europe/Berlin` to my `.config/khal/config`. Googling a little seems to point at a too-new tzlocal - I am indeed running on tzlocal 3.0
> 
> ```
> Traceback (most recent call last):
>   File "/usr/bin/ikhal", line 33, in <module>
>     sys.exit(load_entry_point('khal==0.10.4', 'console_scripts', 'ikhal')())
>   File "/usr/lib/python3.9/site-packages/click/core.py", line 1137, in __call__
>     return self.main(*args, **kwargs)
>   File "/usr/lib/python3.9/site-packages/click/core.py", line 1062, in main
>     rv = self.invoke(ctx)
>   File "/usr/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/usr/lib/python3.9/site-packages/click/core.py", line 763, in invoke
>     return __callback(*args, **kwargs)
>   File "/usr/lib/python3.9/site-packages/click/decorators.py", line 26, in new_func
>     return f(get_current_context(), *args, **kwargs)
>   File "/usr/lib/python3.9/site-packages/khal/cli.py", line 485, in interactive_cli
>     controllers.interactive(
>   File "/usr/lib/python3.9/site-packages/khal/controllers.py", line 545, in interactive
>     pane = ui.ClassicView(
>   File "/usr/lib/python3.9/site-packages/khal/ui/__init__.py", line 1045, in __init__
>     daywalker = Walker(
>   File "/usr/lib/python3.9/site-packages/khal/ui/__init__.py", line 361, in __init__
>     self.ensure_date(this_date)
>   File "/usr/lib/python3.9/site-packages/khal/ui/__init__.py", line 370, in ensure_date
>     pile = self._get_events(day)
>   File "/usr/lib/python3.9/site-packages/khal/ui/__init__.py", line 480, in _get_events
>     self.events = sorted(self._collection.get_events_on(day))
>   File "/usr/lib/python3.9/site-packages/khal/khalendar/khalendar.py", line 152, in get_events_on
>     localize = self._locale['local_timezone'].localize
> AttributeError: 'zoneinfo.ZoneInfo' object has no attribute 'localize'
> ```
> 
> regards, deep42thought

thanks, adding the local timezone solved the start up issue. Did you use vdirsyncer to sync your CalDav calendar? I am guessing I might need to do it since my khal calendar is blank

**Comment by deep-42-thought** [comment]
> thanks, adding the local timezone solved the start up issue. Did you use vdirsyncer to sync your CalDav calendar? I am guessing I might need to do it since my khal calendar is blank

Yes, I use vdirsyncer to sync it with my caldav server - but I was using it with ikhal (and git) solely, before. So, depending on your needs, it may not strictly be necessary to sync the calendar with vdirsyncer.

**Comment by d7415** [comment]
I was going to look at this "properly", but in light of regebro/tzlocal#117 (linked in #985) I think anything like that may be premature. We could pin to 2.x, but at this point I'm temped to wait for a decision upstream.

**Comment by geier** [comment]
this should be fixed with #1126, please re-open if the issue persists.
